src/Fetcher/TululaFetcher.php

```diff
@@ -0,0 +1,235 @@
+<?php
+
+/*
+ * This file is part of the Starfleet Project.
+ *
+ * (c) Starfleet <msantostefano@jolicode.com>
+ *
+ * For the full copyright and license information,
+ * please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace App\Fetcher;
+
+use App\Entity\Conference;
+use App\Entity\Continent;
+use App\Entity\Tag;
+use App\Repository\ExcludedTagRepository;
+use App\Repository\TagRepository;
+use Psr\Log\LoggerInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+class TululaFetcher implements FetcherInterface
+{
+    private const SOURCE = 'tulula';
+    private const TAG_ALLOW_LIST = [
+        'android',
+        'css',
+        'data',
+        'devops',
+        'dotnet',
+        'facebook',
+        'golang',
+        'html',
+        'ios',
+        'java',
+        'javascript',
+        'nodejs',
+        'php',
+        'python',
+        'react native',
+        'ruby',
+        'rust',
+        'security',
+        'ux',
+        'ai/ml',
+        'sql',
+        'iot',
+    ];
+
+    private LoggerInterface $logger;
+    private TagRepository $tagRepository;
+    private ExcludedTagRepository $excludedTagsRepository;
+    private LocationGuesser $locationGuesser;
+    private HttpClientInterface $client;
+
+    public function __construct(LoggerInterface $logger, LocationGuesser $locationGuesser, HttpClientInterface $client, TagRepository $tagRepository, ExcludedTagRepository $excludedTagRepository)
+    {
+        $this->client = $client;
+        $this->logger = $logger;
+        $this->tagRepository = $tagRepository;
+        $this->locationGuesser = $locationGuesser;
+        $this->excludedTagsRepository = $excludedTagRepository;
+    }
+
+    public function isActive(): bool
+    {
+        return true;
+    }
+
+    public function getUrl(array $params = []): string
+    {
+        return 'https://tulu.la/api/public';
+    }
+
+    public function fetch(): \Generator
+    {
+        $validConferences = [];
+        foreach ($this->queryTululaEvents() as $conference) {
+            foreach ($conference['tags'] as $tag) {
+                $lowerTag = strtolower($tag['name']);
+
+                if (\in_array($lowerTag, self::TAG_ALLOW_LIST)) {
+                    $repositoryTag = $this->tagRepository->findTagByName($lowerTag);
+
+                    if ($repositoryTag && $repositoryTag->isSelected()) {
+                        $validConferences[$lowerTag][] = $conference;
+                        break;
+                    }
+                }
+            }
+        }
+
+        foreach ($validConferences as $tagName => $conferencesByTag) {
+            yield $this->getUrl() => $this->denormalizeConferences($conferencesByTag, $this->tagRepository->findTagByName($tagName));
```

Comment:
Here, and on line 84 you are requesting the same tags over and over.

On this line, this a bit a shame since you already fetch the exact very same tag. Let's store it along with the conference. I would even use a SplObjectStorage here. But looks if it's possible and easy.

**Then** you can add local cache on line 84. You can take inspiration from that:
```php
function getElementById(string $id): ?string
{
    var_dump([__FUNCTION__ => $id]);
    switch ($id) {
        case 'a':
            return 'A';
        case 'b':
            return 'B';
        default:
            return null;
    }
}

$tags = [];
for ($i = 0; $i < 3; $i++) {
    foreach (range('a', 'c') as $id) {
        $tag = array_key_exists($id, $tags) ? $tags[$id] : ($tags[$id] = getElementById($id));
        var_dump([$id => $tag]);
    }
}
```